### PR TITLE
Include license files in built distribution

### LINKED ISSUE #18296: LICENSE file(s) not included in published PyPI package
### Bug report

**Bug summary**

A LICENSE file is not included in published PyPI packages. This makes [`pip-licenses`](https://github.com/raimon49/pip-licenses) unable to find a license file when generating a license report.

**Code for reproduction**

```python
$ pip install pip-licenses matplotlib
$ pip-licenses --from=mixed --format=plain-vertical --with-license-file --no-license-path
...
matplotlib
3.3.1
Python Software Foundation License
UNKNOWN
...
```

**Expected outcome**

`pip-licenses` would find a LICENSE file when it inspects the unpacked wheel and print it rather than printing `UNKNOWN`. Newer versions of `setuptools` look in the root directory of a project for LICENSE files when they build, but in this project the LICENSE files are in a subdirectory, so they will need to be explicitly added using the `license_files` option to `setuptools.setup`.

### PR DESCRIPTION
**Pull request by johnthagen**
## PR Summary

Closes #18296 

- Configures `wheel` to include LICENSE files
- Validates CI has included LICENSE files when building wheels
- Normalize LICENSE file names
  - Note this is especially important for what was previously named `Solarized.txt` as there is no way for a tool like `pip-licenses` to glob for LICENSE file named that way that doesn't start with `LICENSE`

Note: [`wheel >= 0.32.0` is needed](https://wheel.readthedocs.io/en/stable/news.html) for this option to be honored.

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
Thanks @johnthagen !

**Comment by jklymak** [comment]
I'm no expert on setup tools, so I guess the question is if this is canonical or if `pip-license` should have been looking in the subdirectories?

**Comment by johnthagen** [comment]
@jklymak This brings up a good point, in the source dist (.tar.gz file), the `LICENSE` folder/files are already included because they have been added to `MANIFEST.in`.

But they are not included in any of the wheel distributions. `pip-licenses` looks at the current virtual environment to find the LICENSE files, so if they are not included in the wheel there is no way for `pip-licenses` to find them.

The subfolder is an important note though. I'll check if `pip-licenses` in its current form can handle a folder of LICENSEs if they are included.

**Comment by jklymak** [comment]
Yeah it was a comment from complete ignorance.  It just seems we can't be the only project that has a sub directory?  Not at all objecting to this PR though.

**Comment by johnthagen** [comment]
From what I can tell, after matplotlib is installed, currently the LICENSEs don't end up in `site-packages/matplotlib-3.3.1-py37.egg` where they end up for other packages.

I think that `setuptools`/`wheel` have some default smarts to pull `LICENSE*` files into there when they are in the top directory. Since these are nested inside a folder, they don't get dropped in there by default. This PR should fix that.

I'm going to try to test the branch out with the sdist.

**Comment by johnthagen** [comment]
Well, it looks like perhaps `license_files` doesn't affect source dists: https://stackoverflow.com/questions/9977889/how-to-include-license-file-in-setup-py-script#comment104408651_48691876

So while the LICENSE files are included in the `.tar.gz` file, they aren't installed into `site-packages/matplotlib-3.3.1-py37.dist-info` (because there must be some magic in `setuptools` that looks for LICENSE* in the root folder).

I *think* this should work for wheels though, which is the default install most users will use.

Another option is to put multiple LICENSE files in the root of the project. `opencv-python` does this and they append all of the third party licenses together into a single file: https://github.com/skvark/opencv-python

File: setup.py [hunk 1]
```diff
@@ -249,6 +249,17 @@ def build_extensions(self):
     long_description=Path("README.rst").read_text(encoding="utf-8"),
     long_description_content_type="text/x-rst",
     license="PSF",
+    license_files=[
```
Comment (tacaswell) [inline setup.py:252]: Will this take a glob pattern instead?

**Comment by tacaswell** [comment]
I am :-1: on moving the license files around or merging them into one file.  There are already a bunch of tools out there that expect them to be where they are.  If setuptools / wheel can not cope with this we should push back on that being an upstream bug.

File: setup.py [hunk 2]
```diff
@@ -249,17 +249,7 @@ def build_extensions(self):
...
-        "LICENSE/LICENSE_AMSFONTS",
-        "LICENSE/LICENSE_BAKOMA",
-        "LICENSE/LICENSE_CARLOGO.txt",
-        "LICENSE/LICENSE_COLORBREWER",
-        "LICENSE/LICENSE_QT4_EDITOR",
-        "LICENSE/LICENSE_STIX",
-        "LICENSE/LICENSE_TORICK",
-        "LICENSE/Solarized.txt",
-    ],
+    license_files=list(map(lambda p: str(p), Path("LICENSE").glob('*'))),
```
Comment (jklymak) [inline setup.py:252]: Small style thing: can this just be `[str(p) for p in Path("LICENSE").glob('*')]` ?

**Comment by johnthagen** [comment]
So after some further investigation, I've now realized that this will need to be fixed via `setup.cfg`:

```ini
[metadata]
license_files = LICENSE/*
```

This was the only way that actually got the LICENSE files into the wheel. `setuptools` and `wheel` seem to have a lot of automatic support for looking for `LICENSE` files in the root of the project, which had skewed my understanding when solving this in the past.

The important reading:

- https://wheel.readthedocs.io/en/stable/user_guide.html#including-license-files-in-the-generated-wheel-file
- https://github.com/pypa/setuptools/issues/1636
- https://setuptools.readthedocs.io/en/latest/history.html#v42-0-0

I moved this to a Draft since I don't want to merge this as is as it doesn't actually solve the problem (sorry for not realizing this earlier).

matplotlib has a `setup.cfg.template`, is this used for the official builds pushed to PyPI? i.e. if we add the following line to the template will that get used in the build?

I tested locally on macOS and with the `setup.cfg` changes in place in a file named `setup.cfg` I can confirm the licenses get pulled into the wheel correctly.

File: setup.py (hunk 2 above)
Comment (johnthagen) [reply setup.py:252]: I would agree, nice catch, but per comment below we'll actually need to move this to `setup.cfg` if we want this to be useful and actually used by `wheel`.

**Comment by anntzer** [comment]
In my view the fact that matplotlib reuses setup.cfg for its own purposes, even though it's really here for setuptools (distutils), is a minor but real annoyance (see e.g. the hacky `del sdist.sdist.make_release_tree` at the top of `setup.py`!) so I would be in favor of moving it to a separate `mplsetup.cfg` (or even `mplsetupcfg.py`, hehe).

**Comment by tacaswell** [comment]
Can you also add a line to https://github.com/matplotlib/matplotlib/blob/master/.github/workflows/cibuildwheel.yml to copy the template over when we build the wheels?

**Comment by johnthagen** [comment]
> Can you also add a line to https://github.com/matplotlib/matplotlib/blob/master/.github/workflows/cibuildwheel.yml to copy the template over when we build the wheels?

@tacaswell Done. First time working with GitHub actions though. Is there a way to look at the built wheels for this PR so I can validate the LICENSE files made it in?

**Comment by tacaswell** [comment]
I don't think we build wheels on every PR, but it is probably worth adding a job to travis or azure to build the wheel and check if the licences are in it to make sure we don't break this in the future.  You can also add a check after we run the ciwheel build in the GH actions to not publish wheels without the license.

**Comment by QuLogic** [comment]
Wheels are currently built on `master`, `v\d+.\d+.x` and tags. They are not restricted by fork though, so you can simulate it by pushing to your own `master` branch.

**Comment by johnthagen** [comment]
For reference, the LICENSE files will be in:

```
matplotlib-<VERSION>-<PLATFORM>.whl
  matplotlib-<VERSION>.dist-info
    LICENSE*
    ...
```

**Comment by johnthagen** [comment]
The simplest way I've found to check if the LICENSE files are being added to the wheel is:

```
$ python setup.py bdist_wheel | grep 'adding license file "LICENSE/LICENSE"'
```

This will return success/0 if the LICENSE files were added and failure/1 if not. This method may be a little brittle though as it relies on the stdout of `wheel`.

**Comment by johnthagen** [comment]
Here is a Python script that will check if the LICENSE files are in the `.whl` files in `dist`. Is there a preference on which solution is used and if is so, where the script should live, etc.?

```py
#!/usr/bin/env python3

"""Check that all .whl files in the dist folder have the correct LICENSE files included."""

from pathlib import Path
import sys
import zipfile

EXIT_SUCCESS = 0
EXIT_FAILURE = 1

LICENSE_FILE_NAMES = [
    'LICENSE',
    'LICENSE_AMSFONTS',
    'LICENSE_BAKOMA',
    'LICENSE_CARLOGO.txt',
    'LICENSE_COLORBREWER',
    'LICENSE_QT4_EDITOR',
    'LICENSE_STIX',
    'LICENSE_YORICK',
]


def main() -> int:
    dist_dir = Path(__file__).parent / 'dist'
    for wheel in dist_dir.glob('*.whl'):
        with zipfile.ZipFile(wheel) as f:
            license_file_names = [Path(path).name for path in sorted(f.namelist())
                                  if '.dist-info/LICENSE' in path]
            if license_file_names != LICENSE_FILE_NAMES:
                return EXIT_FAILURE
    return EXIT_SUCCESS


if __name__ == '__main__':
    sys.exit(main())
```

**Comment by QuLogic** [comment]
That script could be put in `ci/`; the `LICENSE_FILE_NAMES` could be generated from the same glob, and I'm not sure there's any need to have a `main` function.

**Comment by johnthagen** [comment]
@QuLogic I have updated the files as requested. GitHub Actions now has a step that validates that LICENSE files has been included in all built wheels.

File: ci/check_wheel_licenses.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+EXIT_SUCCESS = 0
+EXIT_FAILURE = 1
+
+project_dir = Path(__file__).parent.resolve().parent
+dist_dir = project_dir / 'dist'
+license_dir = project_dir / 'LICENSE'
+
+license_file_names = [Path(path).name
+                      for path in sorted(license_dir.glob('*'))]
```
Comment (QuLogic) [inline ci/check_wheel_licenses.py:25]: `Path.glob` returns `Path` objects already:
```suggestion
license_file_names = [path.name for path in sorted(license_dir.glob('*'))]
```

**Comment by johnthagen** [comment]
I applied @QuLogic's latest suggestion, good catch. I believe this PR is ready for final review/merge.

**Pull request merged**